Add SVID label string

bri365 opened the pull request:
Add optional SVID label string to sections 5.1 and 6.1
Remove unneeded JWT messages in X.509 section 5.1

Addresses #156 

Signed-off-by: Brian Martin <bri365@gmail.com>

bri365 committed:
add optional SVID label string; remove unneeded JWT messages in X.509 section

Signed-off-by: Brian Martin <bri365@gmail.com>

azdagron commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -176,21 +182,6 @@ message X509BundlesResponse {
     // Bundles are ASN.1 DER encoded.
     map<string, bytes> bundles = 2;
 }
-
```
Good catch! Thanks.

rturner3 replied:
Heh, I almost just pushed a PR for this myself. I'm glad I double checked the open PRs first :)

azdagron commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -158,6 +158,12 @@ message X509SVID {
 
     // Required. ASN.1 DER encoded X.509 bundle for the trust domain.
     bytes bundle = 4;
+
```
We should also reflect this change in the workloadapi.proto file in this directory.

azdagron commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -280,6 +271,12 @@ message JWTSVID {
 
     // Required. Encoded JWT using JWS Compact Serialization.
     string svid = 2;
+
```
We should also reflect this change in the workloadapi.proto file in this directory.

azdagron reviewed (commented):
Thanks for opening this @bri365. I'm still musing on the language you introduced and will likely have some additional feedback there, but thought I'd leave a few drive by comments :)

bri365 committed:
update workload proto file

Signed-off-by: Brian Martin <bri365@gmail.com>

bri365 committed:
update workload proto file

Signed-off-by: Brian Martin <bri365@gmail.com>

evan2645 committed:
Update svid label/hint text in workload api

Provide explicit examples of how to use
Mandate value uniqueness
Change name to hint
Further constrain maximum size, make it clearer

Signed-off-by: Evan Gilman <egilman@vmware.com>

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -203,7 +194,7 @@ The `X509SVIDRequest` request message is currently empty and is a placeholder fo
 The `X509SVIDResponse` response consists of a mandatory `svids` field, which MUST contain one or more `X509SVID` messages (one for each identity granted to the client). The `crl` and `federated_bundles` fields are optional. 
 
-All fields in the `X509SVID` message are mandatory.
+All fields in the `X509SVID` message are mandatory, with the exception of the `hint` field. When the `hint` field is set, its value MUST be unique amongst the set of returned SVIDs in the `X509SVIDResponse` message.
```
```suggestion
All fields in the `X509SVID` message are mandatory, with the exception of the `hint` field. When the `hint` field is set (i.e. non-empty), its value MUST be unique amongst the set of returned SVIDs in any given `X509SVIDResponse` message.
```

Trying to clarify that the uniqueness applies to a specific instance of `X509SVIDResponse`.

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -227,6 +218,8 @@ It is often the case that a workload doesn’t know what identity it should assu
 
 In order to support the widest variety of use cases, the X.509-SVID Profile supports the issuance of multiple identities, while also defining a default identity. It is expected that workloads which are aware of multiple identities can handle decision making on their own. Workloads which don’t understand how to leverage multiple identities may use the default identity. The default identity is the  ...
 
+Workloads that understand how to use multiple identities may leverage the optional `hint` field, which can be used to disambiguate identities and inform the workload of which identity should be used for what purpose. For example, `internal` and `external` to denote an SVID for internal or external use, respectively. SPIFFE Workload API implementations SHOULD NOT support values of more than 1024 b ...
```
Is there any coordination required between workload and control plane to determine whether a workload "understands" how to use multiple identities?
Should we describe here (or somewhere more appropriate) what happens if a workload doesn't know how to handle a particular `hint` value? Should a workload raise a failure to the control plane? When making a `X509SVIDRequest`, should a workload be able to describe the set of supported `hint` values?

evan2645 replied:
How do we manage the potential misalignment between hint values provided by the control plane and hints that the workload wants or needs?

```suggestion
Workloads that understand how to use multiple identities may leverage the optional `hint` field, which can be used to disambiguate identities and inform the workload of which identity should be used for what purpose. For example, `internal` and `external` to denote an SVID for internal or external use, respectively. SPIFFE Workload API implementations SHOULD NOT support values of more than 1024 bytes in length. The exact value of the `hint` field is an operator choice and is otherwise unconstrained by this specification.

It is the workload's decision on how to handle the absence of an expected hint, or the presence of an unexpected one.
```

Optionally, we can add `supported_hints` field on the request message so control plane can log or otherwise misalignments

evan2645 replied:
```suggestion
Workloads that understand how to use multiple identities may leverage the optional `hint` field, which can be used to disambiguate identities and inform the workload of which identity should be used for what purpose. For example, `internal` and `external` to denote an SVID for internal or external use, respectively. SPIFFE Workload API implementations SHOULD NOT support values of more than 1024 bytes in length. The exact value of the `hint` field is an operator choice and is otherwise unconstrained by this specification.

It is the workload's responsibility to handle the absence of an expected hint, or the presence of an unexpected one (e.g. fail, warn, etc).
```

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -327,7 +326,7 @@ The `JWTSVIDRequest` request message contains a mandatory `audience` field, whic
 The `JWTSVIDResponse` response message consists of a mandatory `svids` field, which MUST contain one or more `JWTSVID` messages.
 
-All fields in the `JWTSVID` message are mandatory.
+All fields in the `JWTSVID` message are mandatory, with the exception of the `hint` field. When the `hint` field is set, its value MUST be unique amongst the set of returned SVIDs in the `JWTSVIDResponse` message.
```
Same suggested edit applies here as above to X509 version.

evan2645 commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -203,7 +194,7 @@ The `X509SVIDRequest` request message is currently empty and is a placeholder fo
 The `X509SVIDResponse` response consists of a mandatory `svids` field, which MUST contain one or more `X509SVID` messages (one for each identity granted to the client). The `crl` and `federated_bundles` fields are optional. 
 
-All fields in the `X509SVID` message are mandatory.
+All fields in the `X509SVID` message are mandatory, with the exception of the `hint` field. When the `hint` field is set, its value MUST be unique amongst the set of returned SVIDs in the `X509SVIDResponse` message.
```
Here is the suggestion we discussed on the last SIG-Spec call. There is some uncertainty around whether or not the uniqueness requirement would do more harm than good. The argument for uniqueness is "we haven't solved the disambiguation problem if the hints aren't unique". The argument against is "it seems like there may be cases where we would want duplicated hint values, and it's not clear if the hint in its current shape is wholly sufficient for all use cases" ... please correct me if this summary is off :)

In the interest of progress, we decided to have a look at the option below. By specifying client behavior in the presence of duplicate `hint` values, we think we can de-risk the decision because it will be safer to relax the requirement in the future if necessary.

```suggestion
All fields in the `X509SVID` message are mandatory, with the exception of the `hint` field. When the `hint` field is set (i.e. non-empty), SPIFFE Workload API servers MUST ensure its value is unique amongst the set of returned SVIDs in any given `X509SVIDResponse` message. In the event that a SPIFFE Workload API client encounters more than one `X509SVID` message with the same `hint` value set, then the first message in the list SHOULD be considered the winner.
```

evan2645 replied:
SHOULD be selected

bri365 committed:
Clarify multiple hint usage

Co-authored-by: Evan Gilman <evan2645@gmail.com>
Signed-off-by: Brian Martin <bri365@gmail.com>

bri365 committed:
clarify hint uniqueness

Signed-off-by: Brian Martin <bri365@gmail.com>

azdagron reviewed (approved):
\o/

evan2645 committed:
Merge branch 'main' into svid-labels
